File: lib/on-time.js
'use strict';

/**
 * Whether the shop keeps its promises: of the finished jobs that had a due
 * date, how many were done by it, and by how many days the rest missed.
 *
 * Lifted from `renderSLASection` in `renderer/analytics.js`. That function
 * had already been fixed once — it counted voided jobs and the shop's own
 * calibration prints as promises — and it still counted `status ===
 * 'completed'` only. `delivered` is PAST completed in Khayt's pipeline, so
 * the jobs a shop had finished AND handed over were the ones missing from its
 * delivery record. Eighth chart with the fault.
 *
 * A promise is a finished job with a due date. It is kept when the day it was
 * finished is on or before the due day, in the SHOP'S calendar: `completedAt`
 * and `deliveredAt` are instants, and comparing an instant's UTC date with a
 * local due date flips on-time to late near midnight. A job with no due date
 * made no promise and is not counted either way.
 *
 * PURE. `rate` is null when nothing was promised — "kept 0% of no promises"
 * is not a record, it is an absence of one.
 */
(function (global) {

  const listOf = (v) => (Array.isArray(v) ? v : []);
  const DONE = new Set(['completed', 'delivered']);
  const DAY = 86400000;

  function localDay(v) {
    if (!v) return null;
    const s = String(v);
    if (/^\d{4}-\d{2}-\d{2}$/.test(s)) return s;
    const d = new Date(s);
    if (Number.isNaN(d.getTime())) return null;
    return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}-${String(d.getDate()).padStart(2, '0')}`;
  }

  /** The local day a job was finished: completion, else delivery, else the day taken. */
  function finishedDay(o) {
    return localDay(o.completedAt) || localDay(o.deliveredAt) || localDay(o.date);
  }

  function daysBetween(fromDay, toDay) {
    const a = Date.parse(fromDay + 'T00:00:00'), b = Date.parse(toDay + 'T00:00:00');
    if (Number.isNaN(a) || Number.isNaN(b)) return null;
    return Math.round((b - a) / DAY);
  }

  /**
   * @param orders  the book's `printLog`
   * @param opts    { countsForBusiness, since?: 'YYYY-MM-DD' — a job's `date` on or after }
   * @returns {{ promised, onTime, late, rate: number|null, avgDelayDays: number|null,
   *             worstDelayDays: number|null, lateJobs: Array<{id, project, dueDate, finishedDay, delayDays}> }}
   */
  function onTime(orders, opts) {
    const o = opts || {};
    const counts = typeof o.countsForBusiness === 'function' ? o.countsForBusiness : () => true;
    let promised = 0, kept = 0;
    const lateJobs = [];
    for (const job of listOf(orders)) {
      if (!job || !DONE.has(job.status) || job.voidedAt || !counts(job)) continue;
      const due = localDay(job.dueDate);
      if (!due) continue;
      if (o.since && (!job.date || String(job.date) < o.since)) continue;
      const done = finishedDay(job);
      if (!done) continue;
      promised += 1;
      if (done <= due) { kept += 1; continue; }
      const delay = daysBetween(due, done);
      lateJobs.push({ id: job.id, project: job.project || '', dueDate: due, finishedDay: done,
                      delayDays: delay === null ? 0 : delay });
    }
    lateJobs.sort((a, b) => b.delayDays - a.delayDays);
    const late = lateJobs.length;
    const totalDelay = lateJobs.reduce((s, j) => s + j.delayDays, 0);
    return {
      promised, onTime: kept, late,
      rate: promised > 0 ? Math.round((kept / promised) * 1000) / 10 : null,
      avgDelayDays: late > 0 ? Math.round((totalDelay / late) * 10) / 10 : null,
      worstDelayDays: late > 0 ? lateJobs[0].delayDays : null,
      lateJobs,
    };
  }

  const api = { onTime, finishedDay, DONE };
  if (typeof module !== 'undefined' && module.exports) module.exports = api;
  global.KhaytOnTime = api;
})(typeof globalThis !== 'undefined' ? globalThis : this);

File: mac/KhaytCore/Sources/KhaytCore/OnTime.swift
import Foundation

/// Whether the shop keeps its promises — ported to Swift.
///
/// `rate` is nil when nothing was promised: "kept 0% of no promises" is not a
/// record, it is the absence of one.
public enum OnTime {

    public struct LateJob: Sendable, Equatable {
        public let id: String
        public let project: String
        public let dueDate: String
        public let finishedDay: String
        public let delayDays: Int
    }

    public struct Record: Sendable, Equatable {
        public let promised: Int
        public let onTime: Int
        public let late: Int
        public let rate: Double?
        public let avgDelayDays: Double?
        public let worstDelayDays: Int?
        public let lateJobs: [LateJob]
    }

    static let done: Set<String> = ["completed", "delivered"]

    /// The local day a value names, as `YYYY-MM-DD`.
    ///
    /// ── AND THE ANCHORING IS NOT THE SAME AS `waste-trend`'S ──────────────
    ///
    /// The test here is `^\d{4}-\d{2}-\d{2}$` — anchored at BOTH ends — so a
    /// stored day is taken as written and a full timestamp is PARSED and read
    /// in local time. `lib/waste-trend.js` anchors only the start and slices,
    /// so the same string is handled differently by the two modules. That
    /// looks like an inconsistency and is not one to fix inside a port: a
    /// timestamp read in local time can legitimately be a different day from
    /// its own first ten characters, and which of those a rule wants is the
    /// rule's business.
    public static func localDay(_ value: JSONValue?) -> String? {
        guard JSSemantics.truthy(value) else { return nil }
        let s = JSSemantics.text(value)
        if isPlainDay(s) { return s }
        guard let ms = JSDate.parse(s) else { return nil }
        return dayString(ms: ms)
    }

    /// Exactly ten characters of `YYYY-MM-DD`, and nothing after them.
    static func isPlainDay(_ s: String) -> Bool {
        s.count == 10 && DateRange.startsWithADay(s)
    }

    static func dayString(ms: Double) -> String {
        let date = Date(timeIntervalSince1970: ms / 1000)
        let shifted = ms / 1000 + Double(TimeZone.current.secondsFromGMT(for: date))
        let civil = LanCalendar.civil(fromDays: Int(floor(shifted / 86_400)))
        return DateRange.pad(civil.0, 4) + "-" + DateRange.pad(civil.1, 2)
            + "-" + DateRange.pad(civil.2, 2)
    }

    /// The day a job was finished: completion, else delivery, else the day taken.
    public static func finishedDay(_ order: JSONValue) -> String? {
        guard case .object(let o) = order else { return nil }
        return localDay(o["completedAt"]) ?? localDay(o["deliveredAt"]) ?? localDay(o["date"])
    }

    /// Whole days between two stored days.
    ///
    /// Both are parsed as LOCAL midnight, which is why the difference is
    /// ROUNDED rather than divided: across a daylight-saving boundary two
    /// midnights are 23 or 25 hours apart, and an integer division would call
    /// that nought days or two.
    public static func daysBetween(_ from: String, _ to: String) -> Int? {
        guard let a = JSDate.parse(from + "T00:00:00"),
              let b = JSDate.parse(to + "T00:00:00") else { return nil }
        return Int(JSSemantics.round((b - a) / 86_400_000))
    }

    /// `since` is a job's own `date` on or after that day.
    public static func record(_ orders: [JSONValue], since: String? = nil,
                              counts: (JSONValue) -> Bool = BusinessScope.countsForBusiness)
    -> Record {
        var promised = 0, kept = 0
        var lateJobs: [LateJob] = []
        for row in orders {
            guard case .object(let job) = row,
                  case .string(let status)? = job["status"], done.contains(status),
                  !JSSemantics.truthy(job["voidedAt"]), counts(row) else { continue }
            guard let due = localDay(job["dueDate"]) else { continue }
            // `if (o.since && …)` — an EMPTY `since` is falsy and applies no
            // filter at all. Treating it as a real cutoff excluded every job
            // with no date of its own, which is the opposite of "no filter".
            if let since, !since.isEmpty {
                // A job with NO date is excluded by a `since`, not included:
                // `!job.date || String(job.date) < since`.
                guard JSSemantics.truthy(job["date"]),
                      JSSemantics.text(job["date"]) >= since else { continue }
            }
            guard let done = finishedDay(row) else { continue }
            promised += 1
            // On the day counts as kept. A job due Thursday and finished
            // Thursday is not late.
            if done <= due { kept += 1; continue }
            lateJobs.append(LateJob(id: JSSemantics.text(job["id"]),
                                    project: JSSemantics.truthy(job["project"])
                                        ? JSSemantics.text(job["project"]) : "",
                                    dueDate: due, finishedDay: done,
                                    delayDays: daysBetween(due, done) ?? 0))
        }
        // Worst first. JavaScript's sort is stable, so equal delays keep the
        // order the book lists them in.
        lateJobs = lateJobs.enumerated().sorted {
            $0.element.delayDays != $1.element.delayDays
                ? $0.element.delayDays > $1.element.delayDays
                : $0.offset < $1.offset
        }.map(\.element)

        let late = lateJobs.count
        let totalDelay = lateJobs.reduce(0) { $0 + $1.delayDays }
        return Record(
            promised: promised, onTime: kept, late: late,
            rate: promised > 0
                ? JSSemantics.round((Double(kept) / Double(promised)) * 1000) / 10 : nil,
            avgDelayDays: late > 0
                ? JSSemantics.round((Double(totalDelay) / Double(late)) * 10) / 10 : nil,
            worstDelayDays: late > 0 ? lateJobs[0].delayDays : nil,
            lateJobs: lateJobs)
    }
}
